### DISCUSSION TITLE
week11 如何取得執行SQL影響資料筆數

### DISCUSSION THREAD

**Discussion by cmtilo** [Lidemy/forum#158] [後端相關問題]
請問各位同學有關註冊頁面帳號重複處理的問題

目標：
在"註冊處理邏輯"，將使用者輸入 username 與資料庫進行比對。
須取得 sql 語法執行後影響資料筆數。
預期結果為回傳1，表示影響1筆資料、帳號重複。

疑問：
用不同方法執行，得到的結果不一樣，想請教原因為何？

寫法一：
```
<?php
  session_start();
  require_once('conn.php');
  require_once('utils.php');
  $username = escape($_POST['username']); // 故意輸入重複的帳號(以下同)
  $password = password_hash(escape($_POST['password']), PASSWORD_DEFAULT);
  if (empty($_POST['username'])||
      empty($_POST['password'])) {
    header('Location: register.php?errCode=1');
    die();
  }
  
  $sql = "SELECT * FROM cmtilo_blog_users WHERE username = ". $username;
  $result = $conn->query($sql);
  echo $conn->affected_rows; // 回傳1，符合預期

  if ($result !== false && $conn->affected_rows > 0) {
    header('Location: register.php?errCode=2');
    die();
  } else {
    $sql = "INSERT INTO cmtilo_blog_users(username, password) VALUES (?,?)";
    $stmt = $conn->prepare($sql);
    $stmt->bind_param('ss', $username, $password);    
    $result = $stmt->execute();    
    if(!$result) {
      echo $conn->errno;
      die($conn->error);
    }
    $_SESSION['username'] = $username;
    header('Location: index.php');
  }
?>
```

寫法二：語法重複部分省略
```
<?php
  $sql = "SELECT * FROM cmtilo_blog_users WHERE username = ?";
  $stmt = $conn->prepare($sql);
  $stmt->bind_param('s', $username);  
  $result = $stmt->execute();
  echo $stmt->affected_rows; // 回傳-1，不符合預期
?>
```

寫法三：語法重複部分省略
```
<?php
  $sql = "SELECT * FROM cmtilo_blog_users WHERE username = ?";
  $stmt = $conn->prepare($sql);
  $stmt->bind_param('s', $username);  
  $result = $stmt->execute();
  echo $stmt->num_rows; // 回傳0，不符合預期
?>
```
其實還試了很多其他方法 fetch、fetchColumn、mysql_num_rows 等都不對。
寫法一雖然看似可行， 但使用者如 username 輸入字串，sql 語法執行會跳錯。

所以想請教要如何成功抓到影響幾筆資料呢...？

**Comment by cwc329** [comment]
寫法一你可以把 sql log 出來看看和其他成功的 select sql 有什麼差別，
或者直接貼到 mysql 執行看看會報什麼錯誤。

**Reply by cmtilo** [reply]
1054
後來研究是因為少了分號的關係，數字沒問題，但是字串就會報錯

**Comment by cwc329** [comment]
affected_rows 是被影響的行，
你可以看看[這個](https://www.php.net/manual/en/mysqli-stmt.affected-rows.php)
~~可以想一下讀取資料會不會影響資料。~~

**Reply by cwc329** [reply]
我仔細看了一下，
文件裡有寫 `or matched by the last statement executed`，
所以看起來 select 還是可以用 affected_rows 的。

不過在 php  prepared statement 的方法中比較特別的是 select sql 的 affected_rows 會在 store_result 被呼叫之前就被定義，
依照他的官方文件，這種狀況會 affected_rows 變成 -1，可以參照上面的連結。
所以想要讓 affected_rows 是真正 matched rows，
你可以想一下要先執行什麼步驟。

FYI
php 的 mysqli, mysqli_stmt 是 class，
目前看不懂滿正常的，
但是如果在 week16 物件導向的部分有理解，
再回來看這些會比較有感覺。

**Reply by cmtilo** [reply]
```
  $sql = "SELECT * FROM cmtilo_blog_users WHERE username = ?";
  $stmt = $conn->prepare($sql);
  $stmt->bind_param('s', $username);  
  $result = $stmt->execute();
  $stmt->store_result();
  echo $stmt->affected_rows; // 回傳1筆
```
可以得到1了!!
這樣寫法對嗎??

**Reply by aszx87410** [reply]
試試看 $stmt->num_rows; ?
因為是用 select，所以用 num_rows 語意上來說比 affected_rows 更好

https://www.php.net/manual/en/mysqli-stmt.num-rows.php

**Reply by cmtilo** [reply]
好的
不過又產生新的疑惑

```
  $sql = "SELECT * FROM cmtilo_blog_users WHERE username = ?";
  $stmt = $conn->prepare($sql);
  $stmt->bind_param('s', $username);  
  $result = $stmt->execute();
  $stmt->store_result();
  echo $stmt->num_rows; //可以得到回傳1
```
但..直接執行query 卻不行了??
```  
  $sql = "SELECT * FROM cmtilo_blog_users WHERE username ='" . $username ."'";
  $result = $conn->query($sql);
  echo $conn->num_rows; //錯誤訊息Warning: Undefined property: mysqli::$num_rows
```
為什麼會這樣子

**Reply by aszx87410** [reply]
因為你沒有把主體是誰弄清楚，`$result = $conn->query($sql);`
$conn 是跟 mysql 的 connection，$result 才是你執行完 query 的結果，因此是 `$result->num_rows`，不是 `$conn->num_rows`

**Reply by cmtilo** [reply]
我知道了!
所以用query去執行的時候，`$result` 是物件；prepared statement 方法執行時，`$result` 是boolean，此時`$stmt`才是執行結果?

**Reply by aszx87410** [reply]
這個你可以看文件來確認，文件上都有寫說每一個 method 的回傳值是什麼，然後做了什麼事情，文件上講的是最準確的

**Reply by cmtilo** [reply]
好的我再看一次
